src/client/dfs/dfs.c

```diff
@@ -1613,6 +1607,99 @@ dfs_readdir(dfs_t *dfs, dfs_obj_t *obj, daos_anchor_t *anchor, uint32_t *nr,
 	return rc;
 }
 
+int
+dfs_lookup_rel(dfs_t *dfs, dfs_obj_t *parent, const char *name, int flags,
+	       dfs_obj_t **_obj, mode_t *mode)
+{
+	dfs_obj_t		*obj;
+	struct dfs_entry	entry = {0};
+	bool			exists;
+	int			daos_mode;
+	int			rc = 0;
+
+	if (dfs == NULL || !dfs->mounted)
+		return -DER_INVAL;
+	if (_obj == NULL)
+		return -DER_INVAL;
+	if (parent == NULL)
+		parent = &dfs->root;
+	else if (!S_ISDIR(parent->mode))
+		return -DER_NOTDIR;
+
+	rc = check_name(name);
+	if (rc) {
+		D_ERROR("Invalid file/dir Name\n");
+		return rc;
+	}
+	rc = check_access(dfs, geteuid(), getegid(), parent->mode, X_OK);
+	if (rc) {
+		D_ERROR("Permission Denied.\n");
+		return rc;
+	}
+
+	daos_mode = get_daos_obj_mode(flags);
+	if (daos_mode == -1) {
+		D_ERROR("Invalid access mode.\n");
+		return -DER_INVAL;
+	}
+
+	D_ALLOC_PTR(obj);
+	if (obj == NULL)
+		return -DER_NOMEM;
+
+	rc = fetch_entry(parent->oh, DAOS_TX_NONE, name, true, &exists,
+			 &entry);
+	if (rc)
+		D_GOTO(err_obj, rc);
+
+	if (!exists) {
+		D_ERROR("Entry %s does not exist\n", name);
+		D_GOTO(err_obj, rc = -DER_NONEXIST);
+	}
+
+	strcpy(obj->name, name);
+	oid_cp(&obj->parent_oid, parent->oid);
+	oid_cp(&obj->oid, entry.oid);
+	obj->mode = entry.mode;
+
+	/** if entry is a file, open the array object and return */
+	if (S_ISREG(entry.mode)) {
+		daos_size_t elem_size, chunk_size;
+
+		rc = daos_array_open(dfs->coh, entry.oid, DAOS_TX_NONE,
+				     daos_mode, &elem_size, &chunk_size,
+				     &obj->oh, NULL);
+		if (rc != 0) {
+			D_ERROR("daos_array_open() failed (%d)\n", rc);
+			D_GOTO(err_obj, rc);
+		}
+		if (elem_size != 1) {
+			D_ERROR("Invalid Byte array elem size (%zu)\n",
+				elem_size);
+			daos_array_close(obj->oh, NULL);
+			D_GOTO(err_obj, rc);
+		}
+	} else if (S_ISLNK(entry.mode)) {
+		obj->value = entry.value;
+	} else {
+		rc = daos_obj_open(dfs->coh, entry.oid, daos_mode, &obj->oh,
+				   NULL);
+		if (rc) {
+			D_ERROR("daos_obj_open() Failed (%d)\n", rc);
+			D_GOTO(err_obj, rc);
+		}
+	}
+
+	if (mode)
+		*mode = obj->mode;
+	return rc;
+
+err_obj:
+	D_FREE(obj);
+	obj = NULL;
```

Comment:
D_FREE() sets obj to NULL already, and it's a stack value anyway so there's no value in this.